### ISSUE TITLE
Supporting unpublished songs in blog posts

### ISSUE THREAD

**Issue by simonv3** [funmusicplace/mirlo#952]
There's been a regular ask to support unpublished songs in blog posts. 

Currently it's possible to insert a song from an existing album into a blog post (as well attach entire albums). But what's currently not possible is to insert a song that is a work in progress, or unpublished, into a blog post. 

I think from a UX perspective, all the user should be able to do is to click on a button in the blog post that's basically "attach song" and directly upload a song. 

What I have some questions about is how that should be stored data-wise. I see two options: 

1. Create a hidden (or public only to the artist) "drafts" TrackGroup that songs linked to in a Post get stored in and then referenced. 
    * **Pros**: this creates the least changes in infrastructure, and we probably get to re-use a lot of the existing upload functionality. It also maintains a central place to manage all songs uploaded (in release management)
    * **Cons**: it feels a bit "messy" and like we're not accurately naming what's happening.
2. Directly associate tracks to posts. 
    * **Pros** This feels semantically like it's more correct.
    * **Cons** I think a pretty big lift code wise. I'm also wondering if that means we should actually separate out "tracks" as separate from "trackGroups" and have a relation table between "tracks" and "TrackGroup" that's "TrackGroupTrack", which would live alongside "TrackGroupPost". The pro of this is that it would mean tracks could be on multiple track groups but I'm not actually sure if that's meaningful. 
    
One of the things I'm also trying to sus out is how easy it is to switch from one of these to the other. Like, can we migrate from one to the other further down the line if it turns out that's necessary. It feels like there'd be more "data loss" if we just embed songs in the way we currently do, and that's making me wondering if we should be storing songs on a blog post anyway (this would be useful to make playlists out of blog posts, for example).

**Comment by crmsnbleyd** [comment]
my vote is for a drafts trackGroup, not really that much of a kludge, and it's how a lot of content management works conceptually

**Comment by simonv3** [comment]
Yeah, that's the way I'm leaning too, with the additional step of creating a PostTrack table that we used as a "playlist" for a trackgroup anyway.

**Comment by simonv3** [comment]
I guess a third option is to just treat each Post its own TrackGroup, and attach a postId to a TrackGroup. This would make it so that any TrackGroup could be considered a playlist. The tricky thing here is the attaching of existing songs to posts, and would that duplicate the song? I feel like that might be a reason against this.

**Comment by samnabi** [comment]
I like the idea of supporting playlists in posts, but that seems like a new feature that's out of scope for this discussion. I think playlists in posts should be able to support not just an artist's own songs, but other mirlo tracks from other artists too. Can a playlist UI be created from an arbitrary list of track IDs?

My vote is also for a drafts TrackGroup. Although maybe it should be called something other than drafts... since these songs are publicly visible to people in posts. Something like `uncategorized` or `unknownAlbum` maybe?

Another consideration: Can an uncategorized track later be moved to a new album? Or will we just have to reupload?

**Comment by simonv3** [comment]
Saw your post with demo tracks @samnabi and it made me think about this again hah. 

> I think playlists in posts should be able to support not just an artist's own songs, but other mirlo tracks from other artists too. 

Agreed! That's a good thing to consider and does feel like it bumps it into a different feature. 

> Can a playlist UI be created from an arbitrary list of track IDs?

Totally, and probably not that much work to take the existing UI for this and do it (it's based on tracks, not on trackgroup).

> Another consideration: Can an uncategorized track later be moved to a new album? Or will we just have to reupload?

This seems like it'd be a good idea, but I think there's some issues here too. But it does seem like to keep this into consideration it would be better for a Post to be linked to a specific trackId (for a playlist). Because if it was linked to a trackgroup things would get messy. 

So it feels like the most future forward thing to do here would be:

* Create a "hidden album" that is the list of full tracks that aren't part of albums. Maybe done by a simple "isHiddenAlbum" toggle (similar to "isDefault" subscription tier). 
* Create a "postTracks" table that links tracks to posts.

Then, when a user uploads a song to a post, upload it to the `isHiddenAlbum` album (create that album if it doesn't exist), and insert into the `postTracks` table.

**Comment by simonv3** [comment]
Just published something that makes this available.

Let me know if you get a chance to try it out!

**Comment by samnabi** [comment]
Just tried it out, and it works with a couple snags:

1) MP3 is not a supported filetype (not sure if that's a decision you want to keep, but just flagging it)

2) Only one song could be added at a time; I have to save the post and refresh the page to add another one

**Comment by simonv3** [comment]
> MP3 is not a supported filetype (not sure if that's a decision you want to keep, but just flagging it)

hmm yeah that's a good question. This is the same as for albums (we use the same code) but I wonder if it makes sense to allow mp3s here.

> Only one song could be added at a time; I have to save the post and refresh the page to add another one

That seems like a bug!

**Comment by simonv3** [comment]
I can't quite figure out what's going on here, but basically if you add an iframe to a textbox, and then press any key, it will delete the iframe. I think this is a _different_ error that doesn't really have anything to do with bulk uploading tracks, which seems to work fine up to the point where we try to add an iframe, at which point it only seems to place the last iframe.

For the sake of this feature I've limited the uploader to only allow one upload, though I added a bunch of code so that if we can figure that out we can enable that pretty quickly.

I'll open an issue to capture this other error.
